### DISCUSSION TITLE
SVG to MP Filter

### DISCUSSION THREAD

**Discussion by Cicorione** [speedata/publisher#637] [Q&A]
Hi @pgundlach 

I am still dealing with `MetaPost`, I barely grasped `Publisher` and learning `MetaPost` to draw a background is driving me crazy... 🤯

On stack **TeX Stack Exchange** I asked for any utility to convert from `SVG` to `MP` and someone replied me this:

> ConTeXt can include and convert SVG graphics to MetaPost. When you add \enabletrackers[metapost.svg.result] at the start of the document you get the generated MetaPost code in the terminal[...]

https://tex.stackexchange.com/questions/731996/any-utility-to-convert-from-svg-pdf-ps-to-mp-metapost

Now I wonder if you believe it would be possible to create a filter for `Publisher` that uses `ConTeXt` (if installed) to convert a `SVG` file into a `MP` one, in a more human-friendly manner.

Something like: 

```
sp --filter svg2mp input.svg output.mp
```

What do you think about it? 🙏

### TIP

I know that SVG files created for instance with **Inkscape** are filled up with a lot of extra tags, to reduce dramatically the size and weight of SVG files I use this utility that does a really great job:

https://github.com/RazrFalcon/svgcleaner/releases

It is still available a version for MacOS too!  😉

**Comment by Cicorione** [comment]
I did some tests, for instance I converted this shape:

![svg2mp](https://github.com/user-attachments/assets/667d1c95-9039-434c-94c4-7de701b99990)

I cleaned up the **Inkscape** file with **Scour** since **SVG-Cleaner** isn't available for `arm64` and compiling `rust` stuff is overly complicated:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<svg width="4mm" height="8mm" version="1.1" viewBox="0 0 4 8" xmlns="http://www.w3.org/2000/svg">
<path d="m0 0s2.0784-0.087653 2 4-2 4-2 4h4v-8z" fill="#f00"/>
</svg>
```

I grabbed the relative part into the `ConTeXt` template, **it is required LMTX ConTeXt:**

```latex
\enabletrackers[metapost.svg.result]

\startbuffer[svgtest]
<svg id="svg21" width="4mm" height="8mm" version="1.1" viewBox="0 0 4 8" xmlns="http://www.w3.org/2000/svg">
 <g id="layer1">
  <path id="path991" d="m0 0s2.0784-0.087653 2 4-2 4-2 4h4v-8z" fill="#f00"/>
 </g>
</svg>
\stopbuffer

\starttext
\includesvgbuffer[svgtest]
\stoptext
```

And in the output you can actually find the conversion:

```sh
metapost        > svg > result graphic:
    draw image (
      fill ((0,0)..controls(0,0)and(2.0784,0.087653)..(2,-4)..controls(1.9216,-8.087653)and(0,-8)..(0,-8)--(4,-8)--(4,0)&&cycle)
        withcolor svggray(0.941)
      ;
    clip currentpicture to (unitsquare xyscaled (4,-8));
    ) ;
```

A little bit convoluted but it created a `MetaPost` description, next step is to see if the conversion likes to **Publisher**… 🤞

**Reply by pgundlach** [reply]
It probably doesn't like the color `svggray(0.941)`. Need to implement that if really necessary (I would have to look up first what this function does).

**Reply by Cicorione** [reply]
Color is the least part I think, for instance I will replace it with the colors I defined using the syntax you defined. I think is reasonable if the colors can be replaced with a place holder and being modified accordingly into the layout file.

But now I can draw (with limitations) most complex shapes in few minutes and have **ConTeXt** converting them, it is a huge help for me! 🤓

**Comment by Cicorione** [comment]
The conversion from `Inkscape/Scour SVG` to `Context/Metapost` is unfortunately unreliable, I am not sure if it is because `Scour` is doing its compression bad, but it is everything off:

![svg2mp](https://github.com/user-attachments/assets/b387bbba-6100-473b-be25-9023b939b593)

I was day-dreaming again... 😭

### Scour SVG (cleaned)

```svg
<?xml version="1.0" encoding="UTF-8"?>
<svg width="12mm" height="8mm" version="1.1" viewBox="0 0 12 8" >
<rect width="12" height="8" fill="#f00" stroke-linecap="round" />
<path d="m0 0v8c1.6569 0 3-1.3431 3-3v-2c0-1.6569-1.3431-3-3-3z" fill="#999" />
</svg>
```

### XML

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="/usr/share/speedata-publisher/schema/layoutschema-en.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="/usr/share/speedata-publisher/schema/layoutschema-en.xsd" type="application/xml" schematypens="http://http://www.w3.org/2001/XMLSchema"?>
<Layout xmlns="urn:speedata.de:2009/publisher/en" xmlns:sd="urn:speedata:2009/publisher/functions/en" xmlns:xi="http://www.w3.org/2001/XInclude">
  <!-- PAGE SETTINGS -->
  <Trace grid="no" textformat="no" objects="no" />
  <Pageformat width="80mm" height="80mm" />
  <LoadFontfile name="Amaranth-BoldItalic" filename="Amaranth-BoldItalic.ttf" mode="harfbuzz" />
  <LoadFontfile name="Oswald-Light" filename="Oswald-Light.ttf" mode="harfbuzz" features="+frac" />
  <LoadFontfile name="Oswald-Bold" filename="Oswald-Bold.ttf" mode="harfbuzz" features="+frac" />
  <DefineFontfamily fontsize="10" leading="12" name="title">
    <Regular fontface="Oswald-Light" />
    <Bold fontface="Oswald-Bold" />
  </DefineFontfamily>
  <DefineFontfamily fontsize="12" leading="10" name="new">
    <Regular fontface="Amaranth-BoldItalic" />
  </DefineFontfamily>
  <DefineColor name="aglgray" model="RGB" r="153" g="153" b="153" />
  <DefineGraphic name="title">
    filldraw box withcolor colors.aglgray ;
    clip currentpicture to roundedsquare(box.width + 10mm, box.height, 3mm);
  </DefineGraphic>
 <DefineGraphic name="svg2mp">
    fill unitsquare xyscaled (12,8) shifted (0,-8) withcolor red;
    fill (0,0) -- (0,-8) .. controls(1.6569,-8) and (3,-6.6569) .. (3,-5) -- (3,-3) .. controls (3,-1.3431) and (1.6569,0) .. (0,0) -- cycle withcolor colors.aglgray;
    clip currentpicture to (unitsquare xyscaled (12,-8))
  </DefineGraphic>
  <!-- first start with a table -->
  <Record element="data">
    <!-- SVG TO MP -->
    <PlaceObject>
      <Table width="20cm" fontfamily="title">
        <Tr minheight="8mm" align="left" valign="middle">
          <Td graphic="title" padding-left="3.2mm" padding-top="1mm" padding-right="0.6mm">
            <Paragraph>
              <B>
                <NoBreak reduce="fontfit">
                  <Value>TITLE TABLE+SVG>MP</Value>
                </NoBreak>
              </B>
            </Paragraph>
          </Td>
          <Td graphic="svg2mp" align="left" padding-right="3mm" padding-left="4.8mm" padding-top="1.3mm">
            <Paragraph color="white" fontfamily="new">
              <Value>NEW</Value>
            </Paragraph>
          </Td>
        </Tr>
      </Table>
    </PlaceObject>
  </Record>
</Layout>
```
